Logic: improve `FindCommand` to handle boolean expressions

lycjackie opened the pull request:
`FindCommandParser#getKeywordsPredicate(...)` creates a predicate that 
does an implicit boolean OR operation when combining predicates.

This may not desirable for our users as they will not be able to have 
tighter search results 
e.g. Finding for module's `name` with both `Programming` and 
`Methodology`.

We aim to improve our `FindCommand` to improve flexibility in finding 
modules easily by treating the user command arguments as a boolean 
expression.

Consider the following example.

  find name/Structures || (code/CS2040C && credits/4)

We aim to tokenize it to:
* name/Structures
* ||
* (
* code/CS2040C
* &&
* credits/4
* )

Afterwards, we apply Shunting-yard algorithm to parse and process the 
arguments accordingly.

However, `StringTokenizer` is insufficient in extracting prefixed
arguments and boolean operators, as we want to keep boolean operators as
tokens. Likewise, `ArgumentTokenizer` is also not suitable, as we do not
want to use whitespace as delimiter for splitting and extracting
arguments.

Therefore, to enable us to improve the `find` command, let's implement:
* a custom `BooleanExpressionTokenizer` to extract prefixed arguments 
  and boolean operators as tokens
* a custom `BooleanExpressionParser`` which handles the parsing of the 
  argument tokens and apply Shunting-yard algorithm
* declare the operators supported (i.e. `&&` for logical AND, || `for` 
  logical OR, () to group expression to be evaluated first
* Refactor `FindCommandParser` to utilise 
  `BooleanExpressionParser#parse(...)`
* implement and modify unit tests for the `find` command to ensure that 
  the expected behaviour is intended

This PR addresses and resolves #65, and will also fix #129.

* [ 1/14] [Logic: create `BooleanExpressionParser` class](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/bbe2af9b0876b36196c15902904757bc37d4eb43)
* [ 2/14] [Logic: create `Operator` class](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/dc39f75357e97619092dee3f43b669de4cb4a031)
* [ 3/14] [Logic: add helper method on Operator](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/6a1bbaf1847524d9e914d7cc6dd6d4c8f42518b7)
* [ 4/14] [Logic: add helper method on Operator](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/fcc701dbf43fcee9d33f2fe04cb0d38fa8451766)
* [ 5/14] [Logic: add helper method for `BooleanExpressionParser`](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/b2cf5be1f24491605b8d121281771f7b5f7c781f)
* [ 6/14] [Logic: implement `BooleanExpressionParser#parse`](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/165b55b7cfa3996ca5691a6cac443669ab56390f)
* [ 7/14] [Logic: utilise `BooleanExpressionParser#parse`](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/cf8bd29bd127fb14ac0154b81a35e2805b83e8de)
* [ 8/14] [Logic/parser: add and use `BooleanExpressionTokenizer`](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/98431c021497eb62b9d21cd07d4f868118965ad8)
* [ 9/14] [Logic: adopt `CliSyntax` for consistency](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/8d328f5147352dfbf2e7df9c6daf2c5a7a59881b)
* [10/14] [Logic: Improve `BooleanExpressionParser#getKeywordsPredicate`](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/b6b35e2ce08b362c6cb29ca61a4ded2aee291c06)
* [11/14] [FindCommand: update `MESSAGE_USAGE`](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/8e3af821f16a4014e66629c621615723399a2e8b)
* [12/14] [test/parser: update `ParserTest` unit test](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/377aaa302515a13ae01b8cc10522b7b3c9eb7874)
* [13/14] [systemtest/FindCommanSystemTest: fix broken unit tests](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/41dfc7bd2aded1b90d2a894144796a24110c5cfd)
* [14/14] [systemtest/FindCommanSystemTest: add unit tests](https://github.com/CS2113-AY1819S2-T09-1/main/pull/119/commits/39226387202af1fd888ec1faafe0a7c7ccce6567)

Commit:
Logic: create `BooleanExpressionParser` class

Our implementation for `FindCommandParser#getKeywordsPredicate` result
an implicit boolean OR predicate.

This is not desirable for our users as they will not be able to have a
tighter search results. e.g. Finding for module's `name` with both
`Programming` and `Methodology`.

With the aim to improve our `FindCommand`

let's create a new class skeleton `BooleanExpressionParser` to help
us evaluate the argument which the user provide.

Commit:
Logic: create `Operator` class

Our implementation for `FindCommandParser#getKeywordsPredicate` result
an implicit boolean OR predicate.

This is not desirable for our users as they will not be able to have a
tighter search results. e.g. Finding for module's `name` with both
`Programming` and `Methodology`.

Let's create an enum class on all the operators that will be used by
`BooleanExpressionParser` and include:
- precedence to indicate which operators should be evaluated first.

Commit:
Logic: add helper method on Operator

To ease our future implementation.

Let's add helper method to convert `String` to `Operator` and:
- check which operator has higher precedence

Commit:
Logic: add helper method for `BooleanExpressionParser`

To ease our future implementation, we will need a way to parse
arguments into `Predicate`.

Let's create a helper method to do it.

Commit:
Logic: add helper method on Operator

To ease our future implementation.

Let's add helper method to evaluate the operators

Commit:
Logic: implement `BooleanExpressionParser#parse`

Our implementation for `FindCommandParser#getKeywordsPredicate` result
an implicit boolean OR predicate.

This is not desirable for our users as they will not be able to have a
tighter search results. e.g. Finding for module's `name` with both
`Programming` and `Methodology`.

Let's improve our parsing method to make use of `Shunting Yard` so we
could detect if `&&` or `||` is used and parse according to what the
user provide.

Commit:
Logic: utilise `BooleanExpressionParser#parse`

To improve our `FindCommandParser`, we will need to utilise the new
method which we created previously

The new method returns us a composite Predicate which we could use to
instantiate `FindCommand`

Let's update `FindCommandParser#parse# to utilise the new method and:
- remove all unused method
- remove all unused imports

Commit:
test/parser: update `ParserTest` unit test

We have updated the `FindCommandParser` to allow specific boolean logic
in our `FindCommand`

Let's update our unit tests to ensure the behaviour of our parser.

`FindCommandParser` now will return a composite Predicate, as
such, we could not test if two `FindCommand` are equals based on the
predicate alone.

Hence, we will ensure the behaviour of our `FindCommand` in system test
and remove multiple keywords checks for our parser.

Affected file include:
- FindCommandParserTest.java
- AddressBookParserTest.java

Commit:
systemtest/FindCommanSystemTest: fix broken unit tests

We have updated the `FindCommandParser` to allow specific boolean logic
in our `FindCommand`.

These changes broke our system unit tests.

Let's update our system unit tests to fix all our broken unit tests.

Commit:
systemtest/FindCommanSystemTest: add unit tests

We have updated the `FindCommandParser` to allow specific boolean logic
in our `FindCommand`

e.g.
`find name/NAME1 && name/NAME2`
`find code/CODE || ( name/NAME1 && name/NAME2 )

Let's add new system unit tests to ensure the behaviour of
`FindCommand` is what we expect and :
- include negative unit tests

Commit:
FindCommand: update `MESSAGE_USAGE`

The `FindCommand` is able to find expression with boolean expressions.

As such, let's update the `MESSAGE_USAGE` to show appropriate response

Creastery reviewed (changes requested):
Good job implementing Shunting-yard algorithm while utilising OOP concepts as well!
Just a quick remark - 
I noticed you added support for handling parenthesis. 
Could you include that in the find command usage examples as well? 
I understand that due to the way how `ArgumentMultimap` tokenizes the input by spaces, all operators and keywords need to be delimited by whitespace. 
Nevertheless, since you already handle parenthesis, we should include examples of proper usage, and try to tokenize the input better in the next iteration.

Thanks for the effort!

lycjackie commented:
> Good job implementing Shunting-yard algorithm while utilising OOP concepts as well!
> Just a quick remark -
> I noticed you added support for handling parenthesis.
> Could you include that in the find command usage examples as well?
> I understand that due to the way how `ArgumentMultimap` tokenizes the input by spaces, all operators and keywords need to be delimited by whitespace.
> Nevertheless, since you already handle parenthesis, we should include examples of proper usage, and try to tokenize the input better in the next iteration.
> 
> Thanks for the effort!

Alright! 

Although there are some basic handling of the parenthesis, I believe it is not ready for proper usage. 

I will open an issue on this and will refine it further before updating the examples.

Many thanks for the reviews!

Creastery commented:
> Although there are some basic handling of the parenthesis, I believe it is not ready for proper usage.
> 
> I will open an issue on this and will refine it further before updating the examples.

Sure, I agree that the an issue can be created so that we have in sight of this potential enhancement that can be done in v1.4.
However, we still need to update the usage to showcase some potential use cases, and also specify the constraints better so that users are not confused.

lycjackie commented:
> > Although there are some basic handling of the parenthesis, I believe it is not ready for proper usage.
> > I will open an issue on this and will refine it further before updating the examples.
> 
> Sure, I agree that the an issue can be created so that we have in sight of this potential enhancement that can be done in v1.4.
> However, we still need to update the usage to showcase some potential use cases, and also specify the constraints better so that users are not confused.

I have updated the command usage for `FindCommand`. 

As for the constraints would it better to indicate them in our `UserGuide`?

Creastery committed:
Logic/parser: add and use `BooleanExpressionTokenizer`

`BooleanExpressionParser` uses `StringTokenizer` to tokenize the command
arguments.

However, `StringTokenizer` is insufficient in extracting prefixed
arguments and boolean operators, as we want to keep boolean operators as
tokens. Likewise, `ArgumentTokenizer` is also not suitable, as we do not
want to use whitespace as delimiter for splitting and extracting
arguments.

Consider the following example.

  find name/Structures || (code/CS2040C && credits/4)

Both `StringTokenizer` and `ArgumentTokenizer` are unable to split the
tokens into:
* name/Structures
* ||
* (
* code/CS2040C
* &&
* credits/4
* )

But, we want to be able to extract the tokens as above to enable
Shunting-yard algorithm to work nicely without any additional processing
of arguments in `BooleanExpressionParser`.

To enable this, let's create `BooleanExpressionTokenizer`, a string
tokenizer class that extracts all arguments with the specified prefixes
and boolean operators as tokens, and replace the use of
`StringTokenizer` to `BooleanExpressionTokenizer` in
`BooleanExpressionParser`.

Creastery commented:
> 
> I have updated the command usage for `FindCommand`.
> 
> As for the constraints would it better to indicate them in our `UserGuide`?

I think the current constraints you have included now are okay. We can refine this further in v1.4 to make it more user-friendly.

As discussed, there was a need to create a custom tokenizer (`BooleanExpressionTokenizer`) despite already having `ArgumentTokenizer` and `StringTokenizer` to reduce unnecessary and additional overhead in parsing and processing the arguments and operators as tokens in `BooleanExpressionParser`. I hope the `BooleanExpressionTokenizer` class I wrote is robust enough for your needs. 
Let me know if you need any advice extending its functionality further.

Besides that, the necessary changes to the `Name` class (in PR #135) is already merged, so I don't suppose there is any further issues preventing parsing of operators (i.e. `(`, `)`, `&`, `|`) by `BooleanExpressionParser` properly. I'll stress test the `find` command a bit more before approving.

Creastery commented on src/main/java/seedu/address/logic/commands/FindCommand.java:
```diff
@@ -20,14 +22,25 @@
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all modules whose names, code or credits"
-            + " contain any of the specified keywords (case-insensitive)"
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all modules which satisfies the expression"
```
Perhaps it would be better to phrase this as:
"find: Finds all modules which satisfy the expression of search conditions specified and displays them as a list with index numbers."

lycjackie commented:
> > I have updated the command usage for `FindCommand`.
> > As for the constraints would it better to indicate them in our `UserGuide`?
> 
> I think the current constraints you have included now are okay. We can refine this further in v1.4 to make it more user-friendly.
> 
> As discussed, there was a need to create a custom tokenizer (`BooleanExpressionTokenizer`) despite already having `ArgumentTokenizer` and `StringTokenizer` to reduce unnecessary and additional overhead in parsing and processing the arguments and operators as tokens in `BooleanExpressionParser`. I hope the `BooleanExpressionTokenizer` class I wrote is robust enough for your needs.
> Let me know if you need any advice extending its functionality further.
> 
> Besides that, the necessary changes to the `Name` class (in PR #135) is already merged, so I don't suppose there is any further issues preventing parsing of operators (i.e. `(`, `)`, `&`, `|`) by `BooleanExpressionParser` properly. I'll stress test the `find` command a bit more before approving.

Many thanks for your `BooleanExpressionTokenizer`, it was very easy to use and 
the implementation is quite similar to `StringTokenizer` 👍 

I suppose this also resolves #129, where explicit whitespaces is needed to separate operators

Creastery commented on src/main/java/seedu/address/logic/commands/FindCommand.java:
```diff
@@ -20,14 +22,25 @@
+            + "OPERATOR "
+            + "[" + PREFIX_CREDITS + "CREDITS]\n"
+            + "Operator could either be " + OPERATOR_AND + " or " + OPERATOR_OR + "\n"
+            + "Note: you can also use parenthesis to indicate which expression to evaluate first.\n"
```
Perhaps consider rephrasing the above two lines as such:
```
You can chain different search condition together using operators:
OPERATOR "&&" for logical "AND" operation (both conditions A AND B must match)
OPERATOR "||" for logical "OR" operation (either conditions A OR B must match)
You can also use parenthesis to group what search conditions to evaluate first.
```

Creastery commented:
> Many thanks for your `BooleanExpressionTokenizer`, it was very easy to use and
> the implementation is quite similar to `StringTokenizer` 👍

No problem! Good to hear that it works well :)

> I suppose this also resolves #129, where explicit whitespaces is needed to separate operators

Yes, merging this PR will also resolve and fix #129.

Commit:
Logic: adopt `CliSyntax` for consistency

Most of the operators have been moved to `CliSyntax`

Let's remove the duplicated code from `BooleanExpressionParser` to make
our code quality better.

Commit:
Logic: Improve `BooleanExpressionParser#getKeywordsPredicate`

With the implementation of Shunting Yard algorithm, we do not need to
parse multiple keywords into its predicate.

Let's improve on our creation of predicate by doing the following:

If the search term is just a single word.
We will find if the keyword is contain within its particular attributes.

If the search term is a sentence.
We wll find if the keyword matches the sentence exactly.

and:
- add new unit tests for method added.
- remove unused method.

Creastery reviewed (approved):
LGTM. Approving and merging.
Thanks for making the necessary changes and implementing additional unit tests to ensure that the functionality of the feature works as intended.
